plugin/dnstap/handler.go

```diff
@@ -27,17 +27,17 @@ type Dnstap struct {
 	ExtraFormat       string
 }
 
-// TapMessage sends the message m to the dnstap interface.
+// TapMessage sends the message m to the dnstap interface, without interpreting the "Extra" field using metadata.
 func (h Dnstap) TapMessage(m *tap.Message) {
-	h.TapMessageWithMetadata(m, nil, nil, nil)
+	h.TapMessageWithMetadata(m, nil, request.Request{}, nil)
 }
 
 // TapMessageWithMetadata sends the message m to the dnstap interface, with "Extra" field being interpreted by the provided metadata context.
-func (h Dnstap) TapMessageWithMetadata(m *tap.Message, ctx context.Context, state *request.Request, rr *dnstest.Recorder) {
+func (h Dnstap) TapMessageWithMetadata(m *tap.Message, ctx context.Context, state request.Request, rr *dnstest.Recorder) {
```

Comment:
I would suggest to remove the `rr *dnstest.Recorder` at all, and always pass `nil` to `Replacer`. I believe this parameter creates more inconveniences than benefits.

The recorder parameter is needed only for handling several standard labels like
https://github.com/coredns/coredns/blob/b95bc0a7cac365b79d46288e1433a5c5ef903e02/plugin/pkg/replacer/replacer.go#L87-L111

This code can safely work with `nil`, i.e. it will not panic.

The reason why it's not so important to support these DNS response labels (as well as other DNS labels) is because DNSTAP already provides an alternative  standardized way for transferring this data.

Regarding parameter `state`, this is more commonly used data type, and we can keep it. This should not be a big problem for developers of other plugins to provide it. However, if we decide to not support the standard DNS labels (as explained above), we can eliminate this parameter as well. You can pass empty `State{}` to the replacer  (this needs to be tested with all standard labels to avoid panics)

The only important parameter here is Context, which provide access to metadata.

I would appreciate a feedback from community regarding support of standard DNS labels for DNSTAP extra section, i.e.
https://github.com/coredns/coredns/blob/b95bc0a7cac365b79d46288e1433a5c5ef903e02/plugin/pkg/replacer/replacer.go#L37-L57

Reply:
Thanks! I have updated these codes in my new commits. I will update the unit tests tomorrow.

For the `state` parameter, maybe we can just keep the function as:
```go
func (h Dnstap) TapMessageWithMetadata(ctx context.Context, m *tap.Message, state request.Request) 
```
Leave the choice to plugin developers, if they think their plugins does not need to support these standard labels, they simply just pass the empty `State{}`